### PR TITLE
test: improve integration tests

### PR DESCRIPTION
**Pull request by 0xAustrian**
Closes BAL-105

### CODE REVIEW

File: test/integration/BCowPool.t.sol [hunk 1]
```diff
@@ -60,6 +60,8 @@ contract BCowPoolIntegrationTest is Test, BCoWConst, BMath, GasSnapshot {
   }
 
   function testBCowPoolSwap() public {
+    vm.skip(true);
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:63]: why?

File: test/integration/PoolSwap.t.sol [hunk 2]
```diff
@@ -1,92 +1,204 @@
...
+import {GPv2Interaction} from '@cowprotocol/libraries/GPv2Interaction.sol';
+import {GPv2Order} from '@cowprotocol/libraries/GPv2Order.sol';
+import {GPv2Trade} from '@cowprotocol/libraries/GPv2Trade.sol';
+import {GPv2Signing} from '@cowprotocol/mixins/GPv2Signing.sol';
+import {BCoWConst} from 'contracts/BCoWConst.sol';
+import {BCoWPool} from 'contracts/BCoWPool.sol';
+import {GasSnapshot} from 'forge-gas-snapshot/GasSnapshot.sol';
+import {Test, Vm} from 'forge-std/Test.sol';
+import {IBCoWPool} from 'interfaces/IBCoWPool.sol';
+import {ISettlement} from 'interfaces/ISettlement.sol';
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:14]: why do we need this here?

File: test/integration/PoolSwap.t.sol [hunk 3]
```diff
@@ -1,92 +1,204 @@
...
-    pool = factory.newBPool();
-
-    tokenA.approve(address(pool), type(uint256).max);
-    tokenB.approve(address(pool), type(uint256).max);
-
-    pool.bind(address(tokenA), 1e18, 2e18); // 20% weight?
-    pool.bind(address(tokenB), 1e18, 8e18); // 80%
-
+    // TODO: deploy using BCoWFactory
+    pool = new BCoWPool(address(settlement));
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:45]: in this file we should be testing `BPool`
in the other file `BCowPool.t.sol` we should override `BPool` delpoyment for a `BCoWPool`

File: test/integration/PoolSwap.t.sol [hunk 4]
```diff
@@ -1,92 +1,204 @@
...
+
+    dai.approve(address(pool), type(uint256).max);
+    weth.approve(address(pool), type(uint256).max);
+    pool.bind(address(dai), ONE_UNIT, 2e18); // 20% weight?
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:49]: rm `?`

File: test/integration/PoolSwap.t.sol [hunk 5]
```diff
@@ -1,92 +1,204 @@
...
+    // finalize
     pool.finalize();
-    vm.stopPrank();
+    // enable trading
+    pool.enableTrading(APP_DATA);
+
+    // clean dai and weth from the settlement
+    vm.startPrank(address(settlement));
+    dai.transfer(address(0), dai.balanceOf(address(settlement)));
+    weth.transfer(address(0), weth.balanceOf(address(settlement)));
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:59]: hmmmm ok i'd guess, but add it on the BCoWPool test file, not in this one 🙏

File: test/integration/PoolSwap.t.sol [hunk 6]
```diff
@@ -1,92 +1,204 @@
...
-    vm.stopPrank();
+  using GPv2Order for GPv2Order.Data;
 
-    vm.startPrank(swapper);
-    // swap 0.5 tokenA for tokenB
-    tokenA.transfer(address(pool), 0.5e18);
-    tokenB.transferFrom(address(pool), address(swapper), 0.096397921069149814e18);
-    vm.stopPrank();
+  function _makeSwap() internal override {
+    uint32 latestValidTimestamp = uint32(block.timestamp) + MAX_ORDER_DURATION - 1;
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:100]: can move this whole contract to BCoWPool test file

File: test/integration/PoolSwap.t.sol [hunk 7]
```diff
@@ -1,92 +1,81 @@
...
+import {IERC20} from '@cowprotocol/interfaces/IERC20.sol';
 import {BFactory} from 'contracts/BFactory.sol';
-import {IBPool} from 'interfaces/IBPool.sol';
-
 import {GasSnapshot} from 'forge-gas-snapshot/GasSnapshot.sol';
+import {Test, Vm} from 'forge-std/Test.sol';
+import {IBFactory} from 'interfaces/IBFactory.sol';
+import {IBPool} from 'interfaces/IBPool.sol';
 
 abstract contract PoolSwapIntegrationTest is Test, GasSnapshot {
```
Comment (0xAustrian) [inline test/integration/PoolSwap.t.sol:11]: I can move this abstract to another file and rename it to `IntegrationBase` or smth, keeping only relevant stuff for both

File: test/integration/BCowPool.t.sol [hunk 8]
```diff
@@ -2,87 +2,67 @@
...
-      tokenWeightIn: pool.getDenormalizedWeight(address(dai)),
-      tokenBalanceOut: weth.balanceOf(address(pool)),
-      tokenWeightOut: pool.getDenormalizedWeight(address(weth)),
-      tokenAmountIn: buyAmount,
-      swapFee: 0
-    });
+    // clean dai and weth from the settlement
+    vm.startPrank(address(settlement));
+    dai.transfer(address(0), dai.balanceOf(address(settlement)));
+    weth.transfer(address(0), weth.balanceOf(address(settlement)));
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:49]: is this needed? do we check `settlement` balances?

File: test/integration/PoolSwap.t.sol (hunk 7 above)
Comment (wei3erHase) [reply test/integration/PoolSwap.t.sol:11]: hmmm we could do:
- `SwapBase.t.sol`
- `BPoolSwap.t.sol`
- `BCoWPoolSwap.t.sol`

File: test/integration/PoolSwap.t.sol [hunk 9]
```diff
@@ -1,92 +1,81 @@
...
+  IBFactory public factory;
 
-  IERC20 public tokenA;
-  IERC20 public tokenB;
+  IERC20 public dai = IERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F);
+  IERC20 public weth = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
 
-  address public lp = address(420);
-  address public swapper = address(69);
+  address public controller = makeAddr('controller');
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:18]: who's `controller`? what is it for?

File: test/integration/PoolSwap.t.sol [hunk 10]
```diff
@@ -1,92 +1,81 @@
...
+    IBPool(pool).bind(address(weth), ONE_UNIT, 8e18); // 80% weight
+    // finalize
+    IBPool(pool).finalize();
   }
 
   function testSimpleSwap() public {
     _makeSwap();
-    assertEq(tokenA.balanceOf(address(swapper)), 0.5e18);
+    assertEq(dai.balanceOf(swapper.addr), DAI_AMOUNT);
     // NOTE: hardcoded from test result
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:52]: move the hardcoded comment to the constant declaration

File: test/integration/PoolSwap.t.sol [hunk 11]
```diff
@@ -1,92 +1,81 @@
...
-    pool.exitPool(lpBalance, new uint256[](2));
+    uint256 lpBalance = IBPool(pool).balanceOf(lp);
+    IBPool(pool).exitPool(lpBalance, new uint256[](2));
 
     // NOTE: no swap fees involved
-    assertEq(tokenA.balanceOf(address(lp)), 100.5e18); // initial 100 + 0.5 tokenA
+    assertEq(dai.balanceOf(lp), HUNDRED_UNITS + DAI_AMOUNT); // initial 100 + 0.5 dai
     // NOTE: hardcoded from test result
-    assertEq(tokenB.balanceOf(address(lp)), 99.903602078930850186e18); // initial 100 - ~0.09 tokenB
+    assertEq(weth.balanceOf(lp), HUNDRED_UNITS - WETH_AMOUNT); // initial 100 - ~0.09 weth
```
Comment (wei3erHase) [inline test/integration/PoolSwap.t.sol:63]: i'd remove these 2 `NOTE`s

**Review by wei3erHase — COMMENTED** [review]
looking amazing pal!

File: test/integration/BCowPool.t.sol [hunk 12]
```diff
@@ -2,87 +2,67 @@
...
-    IBPool(pool).bind(address(dai), HUNDRED_UNITS, 2e18);
-    IBPool(pool).bind(address(weth), HUNDRED_UNITS, 8e18);
+    deal(address(dai), lp, HUNDRED_UNITS);
+    deal(address(weth), lp, HUNDRED_UNITS);
+
+    deal(address(dai), swapper.addr, ONE_UNIT);
+
+    vm.startPrank(lp);
+    // TODO: deploy using BCoWFactory
+    pool = address(new BCoWPool(address(settlement)));
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:35]: check that @0xteddybear already provided us a BCoWFactory in #94  💘

Comment (0xAustrian) [reply test/integration/BCowPool.t.sol:35]: Nicee

File: test/integration/BCowPool.t.sol (hunk 8 above)
Comment (0xAustrian) [reply test/integration/BCowPool.t.sol:49]: no but the settlement has some dust in DAI and WETH, and dont want it to mess with the tests, just preventive

File: test/integration/PoolSwap.t.sol [hunk 13]
```diff
@@ -1,92 +1,78 @@
...
+  IBFactory public factory;
 
-  IERC20 public tokenA;
-  IERC20 public tokenB;
+  IERC20 public dai = IERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F);
+  IERC20 public weth = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
 
-  address public lp = address(420);
-  address public swapper = address(69);
+  address public lp = makeAddr('lp');
```
Comment (0xteddybear) [inline test/integration/PoolSwap.t.sol:18]: very serious, much professional

File: test/integration/PoolSwap.t.sol [hunk 14]
```diff
@@ -1,92 +1,78 @@
...
-  function setUp() public {
-    tokenA = IERC20(address(deployMockERC20('TokenA', 'TKA', 18)));
-    tokenB = IERC20(address(deployMockERC20('TokenB', 'TKB', 18)));
+  Vm.Wallet swapper = vm.createWallet('swapper');
 
-    deal(address(tokenA), address(lp), 100e18);
-    deal(address(tokenB), address(lp), 100e18);
+  uint256 public constant HUNDRED_UNITS = 100 ether;
+  uint256 public constant ONE_UNIT = 1 ether;
+  // NOTE: hardcoded from test result
```
Comment (0xteddybear) [inline test/integration/PoolSwap.t.sol:24]: If the purpose of these tests is to ensure the contracts work in a real-world scenario and we don't have any errors that are easy to miss by fuzzing (such as confusing buy and sell tokens) I'd much rather have amounts that are both simple to grasp and obvious in their correctness, and would even use ranged asserts after the swap to allow for that

when running an integration test I want to know that the returned DAI from selling 1 WETH

- is _close_ to `3.3k`, and 
- is 100% **not** `1/3.3k`

for that I'd tweak the liquidity to be similar to an existing live balancer pool with the same ratio, and be orders of magnitude greater than the swapped amount to not be bothered by slippage

If we want to for example also make sure the curve is bound more closely to the token with greater weight by doing a swap with more slippage, that should be a separate test

**Review by 0xteddybear — COMMENTED** [review]
code structuring looks solid, some comments on hardcoding different amounts instead. From what I could check they aren't coupled to the settler's existing state, but would love to be corrected on that

File: test/integration/PoolSwap.t.sol (hunk 14 above)
Comment (0xAustrian) [reply test/integration/PoolSwap.t.sol:24]: To be honest I think that is too much for an integration, as you said the purpose of this tests is just test that they work in a real scenario. But lets wait for @wei3erHase opinion on this.

Comment (0xteddybear) [reply test/integration/PoolSwap.t.sol:24]: for clarification, 
> If we want to for example also make sure the curve is bound more closely to the token with greater weight by doing a swap with more slippage, that should be a separate test
would 100% be out of scope for this PR

Comment (wei3erHase) [reply test/integration/PoolSwap.t.sol:24]: current mkt quote is not sth we would wanna replicate (the pool is indistinct to what's current market price), but i agree we could have some more realistic numbers. 

we're deploying a 20-80 pool, which means that the pool will accept that the value should be distributed in that relation. we could make a `1000 DAI = 1 ETH` for simplicity, but then we'll have to replicate it on the pool balances to make it work.

for example, let's say the pool has 1 ETH, and that it represents the 80% of the value, on the other side we should have sth like 250 DAI, if i'm not mistaken.

```
250 * 0,8 = 200
1000 * 0,2 = 200
```

on a similar pool, but 50-50, we should expect to be pairing 1 ETH with 1000 DAI

also agree that out of scope for this PR

File: test/integration/BCowPool.t.sol (hunk 8 above)
Comment (wei3erHase) [reply test/integration/BCowPool.t.sol:52]: hmmm ok but i think is more realistic if we don't do this, anyways, it should be indistinct for the scope of these tests (bc we're not checking settlement balances ever)

File: test/integration/BCowPool.t.sol [hunk 15]
```diff
@@ -173,12 +156,5 @@ contract BCowPoolIntegrationTest is Test, BCoWConst, BMath, GasSnapshot {
     snapStart('settle');
     settlement.settle(tokens, clearingPrices, trades, interactions);
     snapEnd();
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:158]: could you rename it to sth more descriptive?

**Review by wei3erHase — DISMISSED** [review]
small comments not important, can take it from there.
please fix merge conflicts and would be ready to be merged 👍

**Review by wei3erHase — APPROVED** [review]
👍

**Pull request merged**